include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * differentation computations, but can make performing substutitions in
+       * other post-processing computations more straight forward.
+       *
+       * @note For taped AD numbers, this operation is only valid outside recording mode.
+       */
+      std::vector<ad_type>
+      get_non_sensitive_dof_values() const;
+
+      //@}
+
+      /**
+       * @name Operations specific to taped mode: Reusing tapes
+       */
+      //@{
+
+      /**
+       * Set the values for the independent variables $\mathbf{X}$, i.e. the
+       * linearization point.
+       *
+       * @param[in] dof_values A vector field associated with local
+       * degree-of-freedom values on the current finite element. These define
+       * the values of all independent variables.
+       *
+       * @note The input value type must correspond to this class's number type.
+       * So if this class is templated on type double then the number type
+       * associated with value must also be of type double.
+       *
+       * @note If the keep flag has been set when start_recording_operations() is
+       * called then the tape is immediately usable after creation, and the
+       * values of the independent variables set by register_dof_values() are
+       * those at which the function is to be evaluated. In this case, a call to
+       * this function is not strictly necessary.
+       */
+      void
+      set_dof_values(const std::vector<scalar_type> &dof_values);
+
+      /**
+       * Set the values for the independent variables $\mathbf{X}$, i.e. the
+       * linearization point.
+       *
+       * @param[in] values A vector field from which the values of all
+       * independent variables is to be extracted.
+       * @param[in] local_dof_indices A vector of degree-of-freedom indices from
+       * which to extract the local degree-of-freedom values.
+       *
+       * @note If the keep flag has been set when start_recording_operations() is
+       * called then the tape is immediately usable after creation, and the
+       * values of the independent variables set by register_dof_values() are
+       * those at which the function is to be evaluated. In this case, a call to
+       * this function is not strictly necessary.
+       */
+      template <typename VectorType>
+      void
+      set_dof_values(
+        const VectorType &                                  values,
+        const std::vector<dealii::types::global_dof_index> &local_dof_indices);
+
+      //@}
+
+      /**
+       * @name Dependent variables
+       */
+      //@{
+
+      /**
+       * Computes the value of the residual vector field
+       * $\mathbf{r}(\mathbf{X})$.
+       *
+       * @return A Vector object with the value for each component of the vector
+       * field evaluated at the point defined by the independent variable
+       * values.
+       */
+      virtual Vector<scalar_type>
+      compute_residual() const = 0;
```

Comment:
I think I understand this class a lot better now that I see these two member functions. But to understand their names and purpose, I think it would be useful to show a common use case in the class documentation -- namely, how one would implement a derived class. Do you happen to have an example?

Reply:
I will be furnishing each one of the most derived helper classes with a full example (see [here in the next PR](https://github.com/dealii/dealii/pull/6977/files#diff-f54538107ff9ecaf63f8d20799fa3ca7R1011), for example). In total you may expect 5 code samples (the snippet in the very base class and one detailed example for each of the four helper classes). Do you still want me to add another example here? Its seems like a bit of overkill to me.

Reply:
Yes, that would be overkill. Just leave a note in the class documentation that derived classes will have code examples that show how to use this interface -- that ought to be enough.